namedb.c

```diff
@@ -701,3 +701,59 @@ namedb_lookup(struct namedb* db,
 	return domain_table_search(
 		db->domains, dname, closest_match, closest_encloser);
 }
+
+void zone_rr_iter_init(struct zone_rr_iter *iter, struct zone *zone)
+{
+	assert(iter != NULL);
+	assert(zone != NULL);
+
+	memset(iter, 0, sizeof(*iter));
+	iter->zone = zone;
+}
+
+rr_type *zone_rr_iter_next(struct zone_rr_iter *iter)
+{
+	assert(iter != NULL);
+	assert(iter->zone != NULL);
+
+	if(iter->index == -1) {
+		assert(iter->domain == NULL);
+		assert(iter->rrset == NULL);
+		return NULL;
+	} else if(iter->rrset == NULL) {
+		/* ensure SOA RR is returned first */
+		assert(iter->domain == NULL);
+		iter->rrset = iter->zone->soa_rrset;
+	}
+
+	while(iter->rrset != NULL) {
+		if(iter->index < iter->rrset->rr_count) {
+			return &iter->rrset->rrs[iter->index++];
+		}
+		iter->index = 0;
+		if(iter->domain == NULL) {
+			assert(iter->rrset == iter->zone->soa_rrset);
+			iter->domain = iter->zone->apex;
+			iter->rrset = iter->domain->rrsets;
+		} else {
+			iter->rrset = iter->rrset->next;
+		}
+		/* ensure SOA RR is not returned again */
+		if(iter->rrset == iter->zone->soa_rrset) {
+			iter->rrset = iter->rrset->next;
+		}
```

Comment:
The rrsets linked list can contain rrsets from other zones.  You have to skip them.  They can exist for the zone apex (parental zone glue, like a DS record), and there can be sub-zones underneath the zone.  Those have to get skipped after the domain_next and dname_is_subdomain checks.  Those zones underneath the parent zone can still also have rrsets for the parent zone, eg. a DS record or glue record addresses from the parent zone.

Reply:
Corrected. Code actually became somewhat easier to read :+1:. Update the test to include a sub-zone so that the it's guaranteed that they're skipped.